### PR TITLE
Add nightfall (more night attacks)

### PR DESCRIPTION
**Pull request by plague006**

### CODE REVIEW

File: map_gen/misc/nightfall.lua [hunk 1]
```diff
@@ -0,0 +1,148 @@
+--[[
+Softmod rewrite of https://mods.factorio.com/mod/Nightfall written by Yehn and used under the MIT license
+Biters in polluted areas become more aggressive at night.
+TODO: Look into triggering existing unit groups to attack in unison with the groups we generate.
+]]--
+
+local Event = require 'utils.event'
+--basic interval for checks
+local timeinterval = 2700 --2700 is 45 seconds at 60 UPS
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: As per Valansch' comment, might be better as he would probably recommend, use one of the closest prime numbers: 2689 or 2749

File: map_gen/misc/nightfall.lua [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+local bases
+local chunklist
+local state
+local lastattack
+local c_index
+
+local function biter_attack()
+	local maxindex = #bases
+	local surface = game.surfaces[1]
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: This binds it to a single surface :(

File: map_gen/misc/nightfall.lua [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+local ATTACKING = 3
+
+local bases
+local chunklist
+local state
+local lastattack
+local c_index
+
+local function biter_attack()
+	local maxindex = #bases
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: This file contains tabs, the rest is indented with 4 spaces, will recommend to keep it 4 spaces for consistency

File: map_gen/misc/nightfall.lua [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	local surface = game.surfaces[1]
+	for i=c_index, c_index+processchunk, 1 do
+		if i > maxindex then
+			-- we're done here
+			state = IDLE
+			break
+		end
+		if math.random() < surface.darkness then
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: This random is executed quite often, can be optimized:
```lua
-- somewhere near the top of the file
local random = math.random

-- usage in the code:
random()
random(1, 10) -- etc
```

File: map_gen/misc/nightfall.lua [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	end
+end
+
+local function shuffle_table( t )
+    assert( t, "shuffle_table() expected a table, got nil" )
+    local iterations = #t
+    local j
+    
+    for i = iterations, 2, -1 do
+        j = math.random(i)
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: `random(i)`

File: map_gen/misc/nightfall.lua [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	local maxindex = #chunklist
+	for i=c_index, c_index+processchunk, 1 do
+		if i > maxindex then
+			-- we're done with the search
+			state = ATTACKING
+			break
+		end
+		if game.surfaces[1].get_pollution(chunklist[i]) > 0.1 then
+			local chunkcoord = chunklist[i]
+			if (game.surfaces[1].count_entities_filtered{area={{chunkcoord.x-16, chunkcoord.y-16},{chunkcoord.x+16, chunkcoord.y+16}},
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: Same as the above comment with `get_pollution`

File: map_gen/misc/nightfall.lua [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+					type = "unit-spawner"}) > 0 then					
+				table.insert(bases,chunkcoord)
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: Can be referenced in the top of your file: `local insert = table.insert`

File: map_gen/misc/nightfall.lua [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+local function on_chunk_generated(event)
+	-- Track when new chunks are generated and add them on.
+	-- NOTE: The game's debug menu can show potentially hundreds of ungenerated chunks
+	-- It's normal for this count to lag behind chunks in the debug menu.
+	if event.surface == game.surfaces[1] then
+		local chunk = {}
+		local coords = event.area.left_top
+		chunk.x = coords.x+16
+		chunk.y = coords.y+16
+		table.insert(chunklist, chunk)
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: `insert(chunklist, chunk)`

File: map_gen/misc/nightfall.lua [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+			game.surfaces[1].print("bases added: " .. tostring(#bases))
+		end
+	end
+end
+
+local function on_chunk_generated(event)
+	-- Track when new chunks are generated and add them on.
+	-- NOTE: The game's debug menu can show potentially hundreds of ungenerated chunks
+	-- It's normal for this count to lag behind chunks in the debug menu.
+	if event.surface == game.surfaces[1] then
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: This limits it to a single surface

File: map_gen/misc/nightfall.lua [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		table.insert(chunklist, chunk)
+	end
+end
+
+local function on_tick(event)
+	if event.tick % timeinterval == 0 then
+		if game.surfaces[1].darkness > 0.5
+			and state == IDLE
+			and event.tick >= lastattack + timeinterval
+			and math.random() > 0.5
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: `random()`

File: map_gen/misc/nightfall.lua [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+	chunklist = {}
+	state = IDLE
+	--prevents attacks from happening too often
+	lastattack = 0
+	c_index=1
+end
+
+Event.add(defines.events.on_chunk_generated, on_chunk_generated)
+Event.add(defines.events.on_tick, on_tick)
+Event.on_init(on_init)
```
Comment (iltar) [inline map_gen/misc/nightfall.lua]: missing last whitespace in the file

**Comment by Valansch** [comment]
Do you have permission of the author of the mod to use it?
Its under MIT license. We are on GPL3. This (i think) means we can not simply use it, as if it has GPL3.

**Comment by plague006** [comment]
MIT is a permissive license and can be included in GPL works, I'll add the full MIT license text to the file to ensure 100% compliance.

I'll tackle the code changes in the near future.

File: map_gen/misc/nightfall.lua (hunk 2 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Yep.

**Comment by Valansch** [comment]
Huh. 
>The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

I read this as: Your entire software most be under MIT if you use this code.

**Comment by plague006** [comment]
Quoting from https://medium.freecodecamp.org/how-open-source-licenses-work-and-how-to-add-them-to-your-projects-34310c3cf94

MIT License

MIT is one of the most permissive licenses. It is also the most popular one. It offers very low protection for the author of the software.

    The source code doesn’t need to be public when a distribution of the software is made.
    Modifications to the software can be release under any license.
    Changes made to the source code may not be documented.
    It offers no explicit position on patent usage.

File: map_gen/misc/nightfall.lua (hunk 1 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Committed, will push later.

File: map_gen/misc/nightfall.lua (hunk 3 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Committed, will push later.

File: map_gen/misc/nightfall.lua (hunk 4 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Committed, will push later.

File: map_gen/misc/nightfall.lua (hunk 7 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Committed, will push later.

File: map_gen/misc/nightfall.lua (hunk 9 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Yep

File: map_gen/misc/nightfall.lua [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+		bases = {}
+	end
+	local maxindex = #chunklist
+	for i=c_index, c_index+processchunk, 1 do
+		if i > maxindex then
+			-- we're done with the search
+			state = ATTACKING
+			break
+		end
+		if game.surfaces[1].get_pollution(chunklist[i]) > 0.1 then
```
[the remark this answers is not in the document]
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Does https://github.com/Valansch/RedMew/pull/251/commits/1ed2b884cb598a4e767e13e27624c6666096f373 resolve this? Not sure if you wanted it defined at the top or inside the function itself.

File: map_gen/misc/nightfall.lua (hunk 6 above)
Comment (plague006) [reply map_gen/misc/nightfall.lua]: Does https://github.com/Valansch/RedMew/pull/251/commits/27063e0eef8618f43716fc7247aa2aad05fe85c4 resolve this? Not sure if you wanted it defined at the top or inside the function itself.

File: map_gen/misc/nightfall.lua (hunk 12 above)
[the remark this answers is not in the document]
Comment (iltar) [reply map_gen/misc/nightfall.lua]: If a variable is a game related object (surface, entity etc), you should avoid defining it at that point. What I mean is something like this:

```lua
local function do_something(chunks)
    -- ...
    -- this calls the LuaGame, then the surfaces, does a look up for 1, then calls the LuaSurface and calls get_pollution, in a loop
    if game.surfaces[1].get_pollution(chunklist[i]) > 0.1 then
        -- ...
    end
    -- ...
end

-- optimized version
local function do_something(chunks)
    -- rule of thumb: if you use it multiple times, put it in a local
    local surface = game.surfaces[1]
    local get_pollution = surface.get_pollution
    -- ...
    -- merely calls the function "get_pollution"
    if get_pollution(chunklist[i]) > 0.1 then
        -- ...
    end
    -- ...
end

-- surface independent version
local function do_something(surface, chunks)
    local get_pollution = surface.get_pollution
    -- ...
    -- merely calls the function "get_pollution"
    if get_pollution(chunklist[i]) > 0.1 then
        -- ...
    end
    -- ...
end
```

The advantage here is that you can make it multi-surface and you avoid a lookup. This is called inversion of control.

**Pull request merged**